locked-nft/contracts/LockedNFT.cdc

```diff
@@ -0,0 +1,131 @@
+        pub let lockedUntil: UInt64
+        pub let nftType: String
+
+        init (id: UInt64, owner: Address, duration: UInt64, nftType: String) {
+            if let lockedToken = LockedNFT.lockedTokens[id] {
+                self.owner = lockedToken.owner
+                self.lockedAt = lockedToken.lockedAt
+                self.lockedUntil = lockedToken.lockedUntil
+                self.nftType = lockedToken.nftType
+            } else {
+                self.owner = owner
+                self.lockedAt = UInt64(getCurrentBlock().timestamp)
+                self.lockedUntil = self.lockedAt + duration
+                self.nftType = nftType
+            }
+        }
+    }
+
+    pub fun getLockedToken(id: UInt64): LockedNFT.LockedData? {
+        return LockedNFT.lockedTokens[id]
+    }
+
+    pub fun canUnlockToken(id: UInt64): Bool {
+        if let lockedToken = LockedNFT.lockedTokens[id] {
+            if lockedToken.lockedUntil < UInt64(getCurrentBlock().timestamp) {
+                return true
+            }
+        }
+
+        return false
+    }
+
+    pub resource interface LockedCollection {
+        pub fun getIDs(): [UInt64]
+        pub fun borrowNFT(id: UInt64): &NonFungibleToken.NFT
+        pub fun lock(token: @NonFungibleToken.NFT, duration: UInt64)
+        pub fun unlock(id: UInt64): @NonFungibleToken.NFT
+    }
+
+    pub resource Collection: LockedCollection {
+        pub var lockedNFTs: @{UInt64: NonFungibleToken.NFT}
+
+        pub fun unlock(id: UInt64): @NonFungibleToken.NFT {
+            pre {
+                LockedNFT.canUnlockToken(
+                    id: id
+                ) == true : "locked duration has not been met"
+            }
+
+            let token <- self.lockedNFTs.remove(key: id) ?? panic("Missing NFT")
+            LockedNFT.lockedTokens.remove(key: id)
+            LockedNFT.totalLockedTokens = LockedNFT.totalLockedTokens - 1
+            emit NFTUnlocked(id: token.id, from: self.owner?.address)
+
+            return <-token
+        }
+
+        pub fun lock(token: @NonFungibleToken.NFT, duration: UInt64) {
+            let id: UInt64 = token.id
+            let oldToken <- self.lockedNFTs[id] <- token
+            let lockedData = LockedNFT.LockedData(
+                id: id,
+                owner: self.owner!.address,
+                duration: duration,
+                nftType: oldToken.getType().identifier
+            )
+            emit NFTLocked(
+                id: id,
+                to: self.owner?.address,
+                lockedAt: lockedData.lockedAt,
+                lockedUntil: lockedData.lockedUntil
+            )
+            LockedNFT.lockedTokens[id] = lockedData
+            LockedNFT.totalLockedTokens = LockedNFT.totalLockedTokens + 1
+
+            destroy oldToken
+        }
+
+        pub fun getIDs(): [UInt64] {
+            return self.lockedNFTs.keys
+        }
+
+        pub fun borrowNFT(id: UInt64): &NonFungibleToken.NFT {
```

Comment:
Do we want to make it possible to still borrow the nft and maybe read metadata from it? If that is the case we may want to also call this function in the borrowNFT in the actual collection

Reply:
I removed this. Thank you Deewai.